### PR TITLE
Move PyTest functionality from PluginRegistry Class to Unit Tests

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] NO REQUIRED ***unit tests*** (explain why not required)
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** (explain why not required)

# Related Issues
fixes NRLMMD-GEOIPS/geoips#426

# Testing Instructions
To run the new unit tests, simply run ``pytest test_plugin_registries.py -v``. This will unit test the plugin registry and give you verbose output of what occurs. All passes should test other than three, which fail (expected failure). This file can be found at geoips/tests/unit_tests/plugin_registries/test_plugin_registries.py.

# Summary
*From GEOIPS#426: 2024-02-02, GeoIPS won't run pip install . w/out [test] target*

GeoIPS currently fails to run if installed by simply calling ``pip install .``. It must,
instead, be installed with pip install .[test]. This is because pytest is used in
geoips/plugin_registry.py but is not included in the base GeoIPS installation
requirements.

To remediate this, we've since moved the pytest validation functionality off of the
PluginRegistry class itself, and moved it into unit tests at
``tests/unit_tests/plugin_registries/test_plugin_registries.py``, under class
PluginRegistryValidator. This removes the need for PyTest to be a GeoIPS dependency,
and is now a Unit Test dependency. If users want to validate their plugin registry, they
will have to manually install pytest to validate from now on.

    - modified: geoips/plugin_registry.py
    - modified: tests/unit_tests/plugin_registries/test_plugin_registries.py

# Output
Shown below is the expected result of running the pytest command listed above.

Screenshot 2024-02-16 at 11 41 01 AM

### CODE REVIEW

**Comment by jsolbrig** [comment]
@evrose54 I think I'd like to talk about this PR a little bit. You did what I asked, but I think we should think about how the tests are structured a little more carefully and whether some of the validation should remain on in plugin_registry.py.

Questions:
- Should some of the validation remain in the plugin registry code?
- Can we just remove the pytest calls from any tests that we move back?
- Can we move a lot of the validation functions into the `test_` functions rather than defining two classes here?

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (biosafetylvl5) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:208]: why is this marked as a pseudo private var?

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (biosafetylvl5) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:215]: If including a comment like this, we should explain what the error means, and why it's fine

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (biosafetylvl5) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:37]: this is a direct copy-paste - no edits

File: docs/source/releases/v1_12_2a0.rst
Comment (biosafetylvl5) [inline docs/source/releases/v1_12_2a0.rst:37]: Your description of why this is needed on slack was very helpful - can you integrate it into the docs somewhere?

```
Just to ensure it's in the correct state. This was initially automatically done when it was created, but since it invoked pytest and that's not a required dependency, I had to move this to a unit-test module instead. All in all, the user doesn't really need to validate their own registry. It should not be created if an error occurs (duplicate plugin names under the same interface, poorly formatted plugin, etc.), however as developers Jeremy wanted the ability to unit test the PluginRegistry to make sure everything was being done as expected.
```

**Review by biosafetylvl5 — CHANGES_REQUESTED** [review]
This is mostly a copy-paste move, and as such doesn't have much new code to review. I asked for some documentation. @jsolbrig I think you should weigh in on whether this is a good implementation. Let me know if you'd like me to review the code more in-depth.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (jsolbrig) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:215]: I think Evan is saying that he isn't subclassing in order to avoid having this error occur. Rather than inheriting from `PluginRegistry`, he is setting it to the class attribute `plugin_registry`.

File: docs/source/releases/v1_12_2a0.rst
Comment (evrose54) [reply docs/source/releases/v1_12_2a0.rst:37]: I can add this to the documentation.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (evrose54) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py]: Changed it to ``default_fpaths`` instead.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (evrose54) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:215]: Added a comment explaining this more.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (biosafetylvl5) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py]: Thanks! :)

File: docs/source/releases/v1_12_2a0.rst
Comment (biosafetylvl5) [reply docs/source/releases/v1_12_2a0.rst:54]: This is wonderful - thank you so much!

**Review by biosafetylvl5 — APPROVED** [review]
Thank you for making the changes I requested @evrose54 ! :) I'll leave it up to @jsolbrig to review the rest.

**Review by jsolbrig — APPROVED** [review]

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (mindyls) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:37]: So from validate_plugin_types through validate_registry_interfaces, this is all a direct copy/paste from geoips/plugin_registry.py with no additional udpates?

These methods were added to a new "PluginRegistryValidator" class in the test_plugin_registries.py test module (with an init added)?

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (mindyls) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:208]: This is the original existing TestPluginRegistry class (it is just showing up as new right here, because git got confused and showed the class TestPluginRegistry as deleted above)?

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (mindyls) [inline tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: This is just updated so now it uses the internal PR Validator class rather than the imported plugin_registry validator form geoips/plugin_registry.py, right?

**Review by mindyls — APPROVED** [review]
Just resolve my comments if they are correct - I was just leaving notes to make sure I understood what was being updated correctly.

Looks good, and I got 0 return.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (jsolbrig) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:37]: Yes, this is right. We discussed leaving them where they were and extracting the pytest portions, but determined that these tests really are only needed at development time and aren't needed at run time so it made sense to move them to the unit tests.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (jsolbrig) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:208]: Yep, that's right. It thought the previous part was a renaming rather than a move.

File: tests/unit_tests/plugin_registries/test_plugin_registries.py
Comment (jsolbrig) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: Correct. The the PluginRegistry class doesn't really do any validation anymore. It has all been moved here.

Now that I think about it, though, we should test what happens now if the plugin registry is corrupted or malformed. Are the resulting errors helpful? If not, can we make them helpful?

@evrose54 What do you think on this?

Comment (evrose54) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: That's probably a good idea. I don't love that the Plugin Registry isn't validated during it's creation, but it will check to make sure that every plugin added is not duplicately named and that every plugin added is valid in its nature. The way I see it we have two options:

1. Run the validation routines when the plugin registry is created, and remove references to pytest
2. Keep the validation routines as a unit test, however run some final validation at the end of the registry's creation to ensure it's in a correct state. 

A side note:
When we load in the registry to use in GeoIPS, via get_plugin, ... it might be smart to perform some validation of the contents it's reading. Users could possible modify the registry files manually and we should check that the registries are in fact valid when loading them in. Currently we're assuming that after they've been created they are valid, and do no validation during runtime.

Comment (jsolbrig) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: I'm debating whether I think this is safe. I'm not sure if I'm comfortable only validating at creation time, but maybe it is okay.

Maybe another option. Would it be possible to separate the validation routines from the pytest calls? Keep some or all of the validation in the original class, but have it raise errors rather than call pytest.xfail, then catch those errors in the unit tests?

I think that I'm good with approving this PR, but I think we should create another issue to think about where validation should actually occur and how to handle unit testing for this. For example, maybe the part that should be getting tested in the unit tests is actually the validation functions themselves.

Comment (evrose54) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: That's a fair point. I'll create an issue that we should discuss where the validation should occur, and what the unit tests should actually test. Once that is done, I'll merge this PR.

Comment (evrose54) [reply tests/unit_tests/plugin_registries/test_plugin_registries.py:238]: Going to merge now, see issue #496 for more information.

**Pull request merged**
